main/file.c

```diff
@@ -138,18 +159,112 @@ static sub_status_t sleep_unlock(void){
 
 esp_now_send_status_t sub_get_last_message_status(void){
 
-    return ESP_NOW_SEND_SUCCESS;
+    return sub_struct.last_message_status;
 }
 
 sub_status_t sub_enable_sleep(void){
     return true;
 }
 
-static sub_status_t send_cb(){
+static void send_cb(const uint8_t *mac, esp_now_send_status_t status){
 
-    return SUB_OK;
+    sub_send_cb_data_t send_data;
+
+    send_data.message_status = status;
+
+    memcpy(send_data.mac, mac, ESP_NOW_ETH_ALEN);
+
+    xQueueSend(sub_struct.send_cb_queue, &send_data, 0);
+    
 }
 
-static sub_status_t recv_cb(){
-    return SUB_OK;
+static void recv_cb(const esp_now_recv_info_t *info, const uint8_t *data, size_t len){
+
+    if(info->src_addr == NULL || data == NULL || len == 0 || len > MAX_DATA_LENGTH){
+        return;
+    }
+
+    sub_recv_cb_data_t recv_data;
+
+    memcpy(recv_data.mac, info->src_addr, ESP_NOW_ETH_ALEN);
+    memcpy(recv_data.data, data, len);
+    recv_data.data_size = len;
+
+    xQueueSend(sub_struct.recv_cb_queue, &recv_data, 0);
+
+}
+
+static void on_send(sub_send_cb_data_t *send_data){
+
+
+    uint64_t current_time_ms = esp_timer_get_time()/1000;
+
+    sub_struct.last_message_status = send_data->message_status;
+
+    if(send_data->message_status == ESP_NOW_SEND_SUCCESS){
+
+        sub_struct.last_tx_time_ms = current_time_ms;
+
+    }
+    else{
+
+        if(current_time_ms - sub_struct.last_tx_time_ms > sub_struct.tx_nack_timeout_ms){
+
+            if(xSemaphoreTake(sub_struct.mutex, portMAX_DELAY) == pdTRUE){
+
+                sub_struct.interval_ms = sub_struct.transmit_periods[SLEEP_MS];
+
+                xSemaphoreGive(sub_struct.mutex);
+
+            }
+        }
+    }
+}
+
+static void on_recive(sub_recv_cb_data_t *recv_data){
+
+    uint8_t obc_mac_addr[ESP_NOW_ETH_ALEN] = OBC_MAC_ADDRESS;
+
+    if(memcmp(recv_data->mac, obc_mac_addr, ESP_NOW_ETH_ALEN)==0){
+
+        if(recv_data->data_size == STATE_MSG_SIZE && recv_data->data[0] < ENUM_MAX){
+
+            if(xSemaphoreTake(sub_struct.mutex, portMAX_DELAY) == pdTRUE){
+
+                sub_struct.interval_ms = sub_struct.transmit_periods[recv_data->data[0]];
+
+                xSemaphoreGive(sub_struct.mutex);
+            }
+        }
+    }
+    else{
+
+        if(sub_struct._on_data_rx == NULL){
+
+            if(recv_data->data_size > MAX_DATA_LENGTH){
+                    
+                    return;
+            }
+
+            recv_cb_cmd_t *aux_data;
+
+            mempcy(aux_data, recv_data->data, recv_data->data_size); 
```

Comment:
Literówka się wkradła :writing_hand:

Reply:
A i w zasadzie te dwa memcpy z if i elsa są tym samym, bo w uni masz tablice uint8_t o rozmiarze struktury przychodzącej. Więc jak raz skopiujsze do zmiennej recv_cb_cmd_t x, to możesz podać później strukture,
x.cmd lub tablice x.raw